### ISSUE TITLE
amp-user-notification: Inappropriate content and focus order

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#31180]
## WCAG Level
Level A

## Priority
High

## Type
Navigation, Structure

## Pages/screens/components affected
* [amp-user-notification](https://amp.dev/documentation/components/amp-user-notification/)

## Description
When content is not presented in a logical order, it can cause confusion for people who are unable to see its intended order.

For people who navigate using a keyboard or keyboard-like device (such as a switch control or a joystick), the focus order of interactive controls in a document should be logical and intuitive.

The `amp-user-notification` component displays a notification that is visually at the bottom of the browser's viewport. Its source order, and as a result the focus order of any focusable elements it contains, is determined purely by the order where the `<amp-user-notification>` component was placed in the page itself, which - considering that the notification is likely to be important/require the user's attention - is not meaningful or coherent.

## User impact
Depending on the overall page structure, and where the `<amp-user-notification>` element has been placed in the document, sighted keyboard users will have a potentially frustrating and confusing experience. They will see the notification prominently displayed, but will have to navigate through parts of the underlying page before reaching any controls in the notification. For instance, in the main [example page for `amp-user-notification`](https://amp.dev/documentation/examples/components/amp-user-notification/?format=websites), the "I accept" control inside the notification is preceded by 142 other tab stops, making it an exceedingly tedious experience for a keyboard user to reach and acknowledge/dismiss the notification.

![Illustration of the tab stops required to reach the 'I agree' control in the example page for the component](https://user-images.githubusercontent.com/54933393/99182150-0ec80a80-272b-11eb-940e-f543ac474f14.png)

For an assistive technology user, the experience is even more confusing. The notification has a `role="alert"`, and the content of the notification is announced by assistive technologies when the page is loaded/when the notification appears. This will be even more disorienting for users, as they will hear the content of the notification (which may prompt them for an action/confirmation), but will then have to read through/navigate the page to actually find the notification itself (and confirm/dismiss it).

## Required solution
While visually the notification may have been placed "out of the way", at the bottom of the viewport, logically it should be the first element in the document that is read, and any focusable elements inside the notification should be the first elements of the page to receive focus.

This solution must be applied to all instances of the issue identified within the test sample, then applied to all other instances of the same issue identified throughout the rest of the website.

## Implementation guidance
We recommend that as part of the scripting/logic that displays the notification, the actual notification container itself be moved in the document structure to the start of the page, as the first element in the `<body>`. As the notification will then be the first element in the page's reading order, it will be unnecessary to mark it with `role="alert"`, since assistive technology users will encounter the notification right away and won't need to be alerted of its presence by having it set to a role with an implicit ARIA live region.

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied to issues identified within the test sample, and to test the rest of the website for instances of the same issue:
1. Open a page with a user notification.
2. Verify that the notification is the first element in the document structure, and that any focusable controls inside the notification are the first elements to receive focus when navigating using the keyboard (`Tab` / `Shift` + `Tab`).

### Definition of done
Complete all of these tasks before closing this issue or indicating it is ready for retest:
* All issues identified within the test sample have been resolved.
* The rest of the website has been tested for the same issue.
* All issues identified throughout the rest of the website have been resolved or filed as new issues.

## Related standards
* [WCAG 2.1 Success Criterion 1.3.2 Meaningful Sequence (Level A)](https://www.w3.org/TR/WCAG21/#meaningful-sequence)
* [WCAG 2.1 Success Criterion 2.4.3 Focus Order (Level A)](https://www.w3.org/TR/WCAG21/#focus-order)

## More information
* [W3C - Placing the interactive elements in an order that follows sequences and relationships within the content](https://www.w3.org/WAI/WCAG21/Techniques/general/G59)

## Test data
Test date: November 2020
